### PR TITLE
Fix rename of default disk cache directory

### PR DESCRIPTION
**Pull request by zhongwuzw**
### New Pull Request Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [ ] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [ ] I have added the required tests to prove the fix/feature I am adding
* [ ] I have updated the documentation (if necessary)
* [ ] I have run the tests and they pass
* [ ] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / reffers to the following issues: ...

### Pull Request Description

1. Maybe from `5.x`? We changed the default disk cache name of directory, we can't just simply do the rename operation, because the older version's directory is still exist (ps. if user upgrade from older to `5.x`), and may contains many images, if we do not handle these images, they would always exists, cannot access, delete or anything.
2. Remove the compiler warnings of `options`.

### CODE REVIEW

**Comment by dreampiggy** [comment]
@zhongwuzw You can still rename the default path as you want. Just keep the cache folder without any issue.

```
SDImageCacheConfig.defaultCacheConfig.namespacePrefix = @"com.hackemist.SDWebImageCache.default";
```

So I don't think we need to do anything like movement, each time you call method in `SDImageCache`

**Comment by zhongwuzw** [comment]
@dreampiggy , nope, we can't expect users to know everything internal. Many users would just use one shot `UIImageView` category method to load image. We can't just simply rename the default directory and leave many garbage files alone on the disk.

File: SDWebImage/SDImageLoader.m [hunk 1]
```diff
@@ -47,13 +47,13 @@
         }
     }
     if (!image) {
-        SDImageCoderOptions *options = @{SDImageCoderDecodeFirstFrameOnly : @(decodeFirstFrame), SDImageCoderDecodeScaleFactor : @(scale)};
+        SDImageCoderOptions *coderOptions = @{SDImageCoderDecodeFirstFrameOnly : @(decodeFirstFrame), SDImageCoderDecodeScaleFactor : @(scale)};
```
Comment (dreampiggy) [inline SDWebImage/SDImageLoader.m]: This change is adoptable. Because this may shadow the `options` arg from the input arg. But this is not related to this PR.

For each PRs, it should contains the most related things. Don't try to bundle something unrelated to a single PR, because this will make it hard to review, and may cause issue if we want to revert.

Comment (bpoplauschi) [reply SDWebImage/SDImageLoader.m]: I agree with this approach as a generic rule of thumb, but I think we can let this one slip :)

**Comment by dreampiggy** [comment]
@zhongwuzw This change, should be already in the migration guide. For 4.x user, they should read it before upgrating, because some changes are totally break change, which not designed to be compatible. And some change may also change the semantic of same API naming. For example, the method in `SDWebImagePrecher` change the concept of prefetch behavior, even the API naming looks the same.

**Comment by bpoplauschi** [comment]
@dreampiggy this is not the case of changing some APIs - which indeed are subject to reading the migration guide and updating the consumer code.
In this case, I think most users will not become aware that the cache path has changed. It is not a big issue since we will just re-download everything, but I prefer we do the migration. It's the right approach for any storage upgrade. If you upgrade a storage (let's say CoreData), you don't expect the user to handle all the migration, you do as much as possible yourself under the hood. Especially here where we will just keep that folder for no usage, just taking up space.

File: SDWebImage/SDImageLoader.m (hunk 1 above)
Comment (zhongwuzw) [reply SDWebImage/SDImageLoader.m]: @dreampiggy 👌 I just think it's trivial, so I do not create a separate `PR`. I'll be careful next time. 😂

**Comment by bpoplauschi** [comment]
- agree that this migration procedure should be limited to `SDDiskCache`
- we need to find a way to easily check and start the migration only then
- any strong storage API needs to handle data format upgrade/migration, so I propose:
  - adding `migrateCacheFromCachePath:toCachePath:` method to `SDDiskCache` protocol and implement it in our `SDDiskCache` class
  - call that method from `SDImageCache` only if the `com.hackemist.SDWebImageCache.default` path exists

**Comment by zhongwuzw** [comment]
@dreampiggy The path you listed in [Migration Guide](https://github.com/rs/SDWebImage/wiki/5.0-Migration-guide#notable-behavior-changes) is wrong. The correct relative path is `Library/Caches/default/com.hackemist.SDImageCache.default`.

I think it's up to how you understand the manager, if you think all related operation should put into `SDDiskCache`, it can, but maybe it still has issues, because we put the business code into the `SDDiskCache`, because `SDDiskCache` is only responsible to do disk operation. We also can handle this, just add a method to move files, and we do business logical in `SDImageCache`.
If you think manager needs to do these things, I think it's also acceptable.

**Comment by dreampiggy** [comment]
@zhongwuzw No. I say, we have a custom disk cache implementation in that folder. So, let assume you move the 4.x original folder to our new path, it will override and put something 4.x MD5-based files. 

But however, our disk cache does not use MD5 hash for key-mapping. It may be a case that a cache key may be hashed to the same value, which should been a cache miss in our hash check, but it accidentilly hit the original files from 4.x folder. Which will cause issue.

**Comment by zhongwuzw** [comment]
@dreampiggy If any user changed `disk` behavior from `4.x` to `5.x`, I think it's him own responsibility to handle these things, maybe he would try delete all directories if he want. I think the `PR` covered the people who just uses one-shot method.

**Comment by dreampiggy** [comment]
@zhongwuzw If you still need to do a disk cache migration. My consideration is that this code should not been placed into `SDImageCache`. Since you say `If any user changed disk behavior from 4.x to 5.x, I think it's him own responsibility to handle these things`.

For SDWebImage, by default it always use `SDDiskCache`, so if the user does not custom any disk cache implementation. This code should be placed into `SDDiskCache`, right ?

Put any disk reading & writing code in `SDImageCache` level is not designed as a good point. The disk operation should happend on the disk cache level.

**Comment by zhongwuzw** [comment]
@dreampiggy Yeah, I can, what's your opinion about @bpoplauschi 's [proposal](https://github.com/rs/SDWebImage/pull/2412#issuecomment-409515647), we can use `dispatch_once` to do one operation. I already add the check logical, so it would do move only necessary, what you think wether we need add a method in `SDDiskCache` protocol?

**Comment by dreampiggy** [comment]
@zhongwuzw Create a new protocol, just for a compatible method, is overcome. Actually people we use custom disk cache, may not use the default cache prefix because this will cause issue. (Like we use a seperate prefix).

So, maybe we can reach a consensus , some adddtional changes should be provided:
+ Make this older path movement available
+ Use `dispatch_once` to avoid when multiple cache initialize at the same time cause the duplicate movement.
+ Move the total code into `SDDiskCache.m` files implementation files. And do some check, like when `currentCachePath == `SDWebImageCache.default`, earily return.
+ Polish the migration guide and let user who need migration notice this changes. Use can decide whether continue to use 4.x cache path or new 5.x cache path.

**Comment by zhongwuzw** [comment]
@dreampiggy @bpoplauschi The reasons why I put only disk related operation to `SDDiskCache` are:
1. We use `makeDiskCachePath` or something to generate disk cache path in `SDImageCache`, other words, `SDImageCache` would tell `SDDiskCache` where to save. I think we still need keep the getter of default cache disk path in `SDImageCache`.
2. Keep `fileManager` in the same `ioQueue`, in `SDDiskCache`, almost all methods is synchronous, I think we shouldn't let the migrate operation in another secondly thread.

What's your guys think about? 🤔

File: SDWebImage/SDImageCache.m [hunk 2]
```diff
@@ -86,7 +86,7 @@ - (nonnull instancetype)initWithNamespace:(nonnull NSString *)ns
         
         NSAssert([config.diskCacheClass conformsToProtocol:@protocol(SDDiskCache)], @"Custom disk cache class must conform to `SDDiskCache` protocol");
         _diskCache = [[config.diskCacheClass alloc] initWithCachePath:_diskCachePath config:_config];
-        [self moveOldDefaultDiskPathOfFilesToNewDefaultDiskPath];
+        [self migrateCacheDiskDirectory];
```
Comment (dreampiggy) [inline SDWebImage/SDImageCache.m]: ...emmm. Am I miss something ?

I means, move all code, into `[SDDiskCache initWithCachePath:config:]` method. Don't need to modify anything in `SDImageCache.m`

During init method of `SDDiskCache`, use dispatch_once + global queue to only move once.

Comment (zhongwuzw) [reply SDWebImage/SDImageCache.m]: @dreampiggy emm, I know you opinion, so I put the reasons in [here](https://github.com/rs/SDWebImage/pull/2412#issuecomment-409621832). I can explain my thoughts further.
Why I put the default cache path in `SDImageCache` is we generate the default cache path in `SDImageCache`, right? Let's see how we generate in `SDImageCache`:

```
// class of SDImageCache
- (instancetype)init {
    return [self initWithNamespace:@"default"];
}
- (nonnull instancetype)initWithNamespace:(nonnull NSString *)ns {
    NSString *path = [self makeDiskCachePath:ns]; // first component disk cache path 
    return [self initWithNamespace:ns diskCacheDirectory:path];
}
- (nonnull instancetype)initWithNamespace:(nonnull NSString *)ns
                       diskCacheDirectory:(nonnull NSString *)directory
                                   config:(nullable SDImageCacheConfig *)config {
    if ((self = [super init])) {
        NSString *namespacePrefix = config.namespacePrefix;
        if (!namespacePrefix) {
            namespacePrefix = @"";
        }
        NSString *fullNamespace = [namespacePrefix stringByAppendingString:ns]; 
      
        // Init the disk cache
        if (directory != nil) {
            _diskCachePath = [directory stringByAppendingPathComponent:fullNamespace];
        } else {
            NSString *path = [self makeDiskCachePath:ns];
            _diskCachePath = path;
        }
    return self;
}
```

We combine many components to generate default cache disk path. The `SDDiskCache` is just a class that do disk operations, like access, delete, move, right? other words, it doesn't need to know what's the disk path it is, its only job is do disk operation, I think compare old and new disk path is not what it should do. If you still think we should put all these things to `SDDiskCache`, should we need to hard code the default path? or extract the methods of default cache path related methods into a single class to let both `SDImageCache` and `SDDiskCache` access it?

**Comment by bpoplauschi** [comment]
I think calling private methods is just a bad practice (Apple discourages this) and error prone.
We can always create a private category over a class that we use internally, but do not expose through our installations. But that is just not needed. We can expose this method on `SDDiskCache`, no problem in it being available.

**Comment by bpoplauschi** [comment]
I just merged #2417 so closing this PR. Thanks @zhongwuzw for driving this conversation, it's great to have the migration code in place.
